keel-docker/src/test/kotlin/com/netflix/spinnaker/keel/docker/SampleDockerImageResolverTests.kt

```diff
@@ -134,27 +153,74 @@ class SampleDockerImageResolverTests : JUnit5Minutests {
         before {
-          every { repository.latestVersionApprovedIn(referenceDeliveryConfig, artifact, "test") } returns null
+          every { repository.latestVersionApprovedIn(referenceDeliveryConfig, artifacts.first(), "test") } returns null
         }
 
         test("the resolver throws an exception") {
-          expectThrows<NoDockerImageSatisfiesConstraints> { this.invoke(versionedContainerResource) }
+          expectThrows<NoDockerImageSatisfiesConstraints> { this.invoke(referenceResource) }
         }
       }
 
       context("there is a version approved") {
         before {
-          every { repository.latestVersionApprovedIn(referenceDeliveryConfig, artifact, "test") } returns "v0.0.1"
+          every { repository.latestVersionApprovedIn(referenceDeliveryConfig, artifacts.first(), "test") } returns "v0.0.1"
         }
 
         test("the resolver finds the correct version") {
-          val resolvedResource = this.invoke(versionedContainerResource)
+          val resolvedResource = this.invoke(referenceResource)
           expect {
             that(resolvedResource.spec.container).isA<DigestProvider>()
             that(resolvedResource.spec.container as DigestProvider).get { digest }.isEqualTo("sha256:2763a2b9d53e529c62b326b7331d1b44aae344be0b79ff64c74559c5c96b76b7")
           }
         }
       }
     }
+
+    context("resolving from multiple reference providers") {
+      before {
+        every { repository.deliveryConfigFor(multiReferenceResource.id) } returns multiReferenceDeliveryConfig
+        every { repository.environmentFor(multiReferenceResource.id) } returns multiReferenceDeliveryConfig.environments.first()
+      }
+
+      context("no versions are approved yet") {
+        before {
+          every { repository.latestVersionApprovedIn(multiReferenceDeliveryConfig, artifacts.elementAt(0), "test") } returns null
+          every { repository.latestVersionApprovedIn(multiReferenceDeliveryConfig, artifacts.elementAt(1), "test") } returns null
+        }
+
+        test("the resolver throws an exception") {
+          expectThrows<NoDockerImageSatisfiesConstraints> { this.invoke(multiReferenceResource) }
+        }
+      }
+
+      context("there is a version approved for each artifact") {
+        before {
+          every { repository.latestVersionApprovedIn(multiReferenceDeliveryConfig, artifacts.elementAt(0), "test") } returns "v0.0.1"
+          every { repository.latestVersionApprovedIn(multiReferenceDeliveryConfig, artifacts.elementAt(1), "test") } returns "v0.0.2"
+        }
+
+        test("the resolver overrides to the latest artifact") {
+          val resolvedResource = this.invoke(multiReferenceResource)
+          expect {
+            that(resolvedResource.spec.container).isA<DigestProvider>()
+            that(resolvedResource.spec.container as DigestProvider).get { digest }.isEqualTo("sha256:b4857d7596462aeb1977e6e5d1e31b20a5b5eecf890cd64ac62f145b3839ee97")
+          }
```

Comment:
Hmm, this test is a bit confusing to me. I was expecting the resolver to translate each of the artifact references to a digest, but the end result here is that the last artifact's digest takes precedence. In a real-world scenario, I'd imagine you'd have a Kubernetes wrapper resource with, say, 2 different `PodSpec` objects each pointing to a different artifact. My guess is you'll override `updateContainerInSpec` for that purpose, but because the resource spec here contains a single `container` field, the test looks a bit weird. If nothing else, maybe just add a comment explaining why?

Reply:
Thanks for the comments @luispollo. The name change looks good to me. will update the PR.

As for the test, yeah your guess is spot on. The k8s plugin will override different container images in the PodSpec by cross-referencing them against artifact references. I have a full-blown set of tests in the k8s plugin ([see here](https://github.com/nimakaviani/managed-delivery-k8s-plugin/blob/f6fefb02a0105a9f49f382e7497f9fbd4a837f49/k8s-plugin/src/test/kotlin/com/amazon/spinnaker/keel/tests/DockerImageResolverTest.kt)) that exercise different permutations. Particularly for multiple containers in the spec [it verifies that the references get written correctly](https://github.com/nimakaviani/managed-delivery-k8s-plugin/blob/f6fefb02a0105a9f49f382e7497f9fbd4a837f49/k8s-plugin/src/test/kotlin/com/amazon/spinnaker/keel/tests/DockerImageResolverTest.kt#L171-L219). 

Because of the exhaustive testing in the plugin, I kept it relatively minimal in the Keel test. I didn't want to duplicate the testing behavior or make significant changes to the existing Sample Resolver; hence just verifying that the resource goes through the DockerImageResolver twice and that eventually the value for the container gets overriden by the second artifact that comes through. 

If you think the explanation is sufficient, I can leave it only as a comment in the test. If it needs further clarification, I'd be happy to make the necessary changes where needed.

Reply:
Makes sense, thanks for the details!